### PR TITLE
Improve codec unsupported detection

### PR DESCRIPTION
**Pull request by hicom150**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix
[ ] New feature
[ ] Other, please explain:

**What changes did you make? (Give an overview)**
It seems that from time to time we get some "codec unsupported" false positive. This PR changes unsupported codec detection to use  `HTMLMediaElement.play() Promise` based on what it is explained in [this](https://developers.google.com/web/updates/2016/03/play-returns-promise) article.

**Is there anything you'd like reviewers to focus on?**
Check that we don't get false positives caused by timing issues.

### CODE REVIEW

**Comment by feross** [comment]
Wow, can it really be this simple? If this actually works better, let's merge it. Haven't had time to test it.

File: src/renderer/pages/player-page.js [hunk 1]
```diff
@@ -56,6 +54,14 @@ function renderMedia (state) {
       mediaElement.pause()
     } else if (!state.playing.isPaused && mediaElement.paused) {
       mediaElement.play()
+        .then(() => {
+          dispatch('mediaSuccess')
+        })
+        .catch((err) => {
```
Comment (Borewit) [inline src/renderer/pages/player-page.js]: Maybe better to comply with: _[use promise then() instead of catch() with callbacks](https://github.com/webtorrent/webtorrent-desktop/pull/1679)_, to prevent we handle exception of the `dispatch` as pointed out by master @feross 😉.

**Comment by Borewit** [comment]
Great work @hicom150.

File: src/renderer/pages/player-page.js [hunk 2]
```diff
@@ -56,6 +54,14 @@ function renderMedia (state) {
...
+          if (err.name === 'NotSupportedError') {
+            dispatch('mediaError', 'Codec unsupported')
+          }
```
Comment (Borewit) [inline src/renderer/pages/player-page.js]: Maybe remove the condition and display any error?

File: src/renderer/pages/player-page.js (hunk 1 above)
Comment (hicom150) [reply src/renderer/pages/player-page.js]: Sure! 👍

File: src/renderer/pages/player-page.js (hunk 2 above)
Comment (hicom150) [reply src/renderer/pages/player-page.js]: Yes! In our case all the road lead to `Codec unsupported` 😅

File: src/renderer/pages/player-page.js [hunk 3]
```diff
@@ -54,14 +54,10 @@ function renderMedia (state) {
...
-          dispatch('mediaSuccess')
-        })
-        .catch((err) => {
-          if (err.name === 'NotSupportedError') {
-            dispatch('mediaError', 'Codec unsupported')
-          }
-        })
+        .then(
+          () => dispatch('mediaSuccess'),
+          () => dispatch('mediaError', 'Codec unsupported')
```
Comment (Borewit) [inline src/renderer/pages/player-page.js]: Maybe?
```js
        .then(
          () => dispatch('mediaSuccess'),
          err => dispatch('mediaError', err.name === 'NotSupportedError' ? 'Codec unsupported' : `${err.name}: ${err.message}`)
        )
```
I put it on a branch: https://github.com/webtorrent/webtorrent-desktop/tree/issue-1711-improve-error-handling

For some reason I would send PR to your clone, I think select your fork as the base. Maybe a bug in GitHub, to many forks of this repo.

Comment (hicom150) [reply src/renderer/pages/player-page.js]: I think that we should dispatch `mediaError` when error type is `NotSupportedError` (original code) or when any error case (like the latest code change).

Given that the possible [exceptions](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play#Exceptions) are `NotSupportedError` or `NotAllowedError` I wouldn't show this last one to the final user 😅

Comment (Borewit) [reply src/renderer/pages/player-page.js]: > I wouldn't show this last one to the final user

why not?

Comment (Borewit) [reply src/renderer/pages/player-page.js]: It also says _other exceptions may be reported,_. I suppose there are not very likely, But it may be good user gets an error if something odd happens.

Comment (Borewit) [reply src/renderer/pages/player-page.js]: With the current code, in the unlikely event of an error named `NotAllowedError`, or any other exception, it will prompt the user with 'Codec unsupported' right?

Comment (hicom150) [reply src/renderer/pages/player-page.js]: > With the current code, in the unlikely event of an error named NotAllowedError, or any other exception, it will prompt the user with 'Codec unsupported' right?

Yes that is right 😅 Although it is not 100% correct for the final user, the action should be the same: **open an external player to be able to play the file**, otherwise it would see the message `The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.` which I found a bit confusing for the user as it doesn't give any "call to action" to solve the problem.

On the other hand, it is true that we should log the proper error code, so we can debug future problems easily 😉

What do you think @Borewit ?

**Comment by RecoX** [comment]
I believe the current implementation using the audio/video api for manage this flows is ok, I tried in the past use this approach but I ended up leaving as is because it seems wiser anyways this pr looks simple.

I would like to add some details I found.

I put a console.log, to see that element as a lot of times is throwing me a false positive
```
  function onCanPlay (e) {
    const elem = e.target
    console.log(666, elem.webkitAudioDecodedByteCount)
    
    if (elem.readyState < HTMLMediaElement.HAVE_FUTURE_DATA) return
    if (state.playing.type === 'video' &&
      elem.webkitVideoDecodedByteCount === 0) {
      dispatch('mediaError', 'Video codec unsupported')
    } else if (elem.webkitAudioDecodedByteCount === 0) {
      dispatch('mediaError', 'Audio codec unsupported')
    } else {
      dispatch('mediaSuccess')
      elem.play()
    }
  }

```
Screen Shot 2019-10-08 at 2 37 04 PM


Both are mp3's so im not sure if this `webkitAudioDecodedByteCount` is really trustable

**Comment by Borewit** [comment]
@RecoX 
> I believe the current implementation using the audio/video api for manage this flow is ok, 

With the current implementation we get false positives as described in #1668 

> I tried in the past use this approach but I ended up leaving

Why? If you had an issue with PR approach, is that still valid? Can you test the PR @RecoX ? Or do you advice to merge, it is not clear to me.

> anyways this pr looks simple

Looks very clean and simple indeed.

**Comment by Borewit** [comment]
As there is some discussion, I suggest to keep the PR open for 2 day. Please use the review process fellow contributors.

Sorry for the delay @hicom150.

**Comment by hicom150** [comment]
> Sorry for the delay @hicom150.

Don't worry @Borewit 😉

I find it super helpful to get feedback from other developers, I think that the code gets better when we discus and combine different point of views.

BTW, thank you for your review and great feedback @Borewit 💪

**Comment by RecoX** [comment]
> @RecoX
> 
> > I believe the current implementation using the audio/video api for manage this flow is ok,
> 
> With the current implementation we get false positives as described in #1668
> 
> > I tried in the past use this approach but I ended up leaving
> 
> Why? If you had an issue with PR approach, is that still valid? Can you test the PR @RecoX ? Or do you advice to merge, it is not clear to me.
> 
> > anyways this pr looks simple
> 
> Looks very clean and simple indeed.

I tested it and it works ok, but I found a problem.

When I tried to reproduce a `wma` it throws the modal with the unsupported codec and that is ok but as you see in the console is doing that in an infinite loop so the program freezes.

https://github.com/refreex/refreex-webtorrent-desktop/blob/master/src/renderer/pages/audio-player.js

here the link of the file where i implement it just in case you want to take a look.

Screen Shot 2019-10-09 at 9 33 43 AM

**Comment by hicom150** [comment]
@RecoX I can't reproduce the error case 😓 but I noticed that you didn't delete [onCanPlay](https://github.com/refreex/refreex-webtorrent-desktop/blob/master/src/renderer/pages/audio-player.js#L116) and [onError](https://github.com/refreex/refreex-webtorrent-desktop/blob/master/src/renderer/pages/audio-player.js#L113) as it is in this PR code 😅

**Comment by RecoX** [comment]
> @RecoX I can't reproduce your error case 😓 but I noticed that you didn't delete [onCanPlay](https://github.com/refreex/refreex-webtorrent-desktop/blob/master/src/renderer/pages/audio-player.js#L116) and [onError](https://github.com/refreex/refreex-webtorrent-desktop/blob/master/src/renderer/pages/audio-player.js#L113) as it is in this PR code 😅

You don't see them because i didn't push the changes, probably is because the player in my fork is in the main window instead of the player window. Anyways is strange that that occurs to me with the new code and with the older no.

**Comment by mathiasvr** [comment]
Video with AC3 audio plays silently (like Chrome), and does not give the user the option to open in an external player.
I also had a false positive (saw the `Codec unsupported` error on a H264 file), which worked fine all other times I tried.

**Comment by hicom150** [comment]
> Video with AC3 audio plays silently (like Chrome), and does not give the user the option to open in an external player.
I also had a false positive (saw the Codec unsupported error on a H264 file), which worked fine all other times I tried.

Ups! I didn't expect that result 😓

In order to help the debugging process can we share the test files where we get the unexpected behaviors? @mathiasvr @RecoX 

Here I share the some files a what should be the expected result:

- [Video: mpeg2video. Audio: ac3](https://send.firefox.com/download/6ade81cef88e9397/#sTLqQZMGQZ5qFaQ612rYrA)
**expected result**: `Codec Unsupported` dialog.
screenshot2

- [Video: h264 Audio: none](https://send.firefox.com/download/c629bf2287e1d275/#uj_xQmUe2CF15elVZR6UAQ)
**expected result**: Plays video normally.
screenshot4

**Comment by mathiasvr** [comment]
@hicom150 I just went ahead and generated a bunch of different samples with various video/audio codecs, because why not? 😃

https://drive.google.com/open?id=1ACw6hOooQNfc-oniyL0Gp-z4OSq-5Mc7

Before this PR, supported a video file (such as h264) with AC3 audio codec displays 'Audio codec not supported', and after this PR it plays the video with audio muted. It would be nice if we could still detect unsupported audio.

**Comment by hicom150** [comment]
After some further testing with the awesome @mathiasvr video smaples 😉 it seems that `HTMLMediaElement.play() Promise` is not as fine grained as the previous implementation 😭

So unless someone comes with a better solution, I think that the only proper option is to try to optimize the previous `onCanPlay` function 😅

**Comment by RecoX** [comment]
> After some further testing with the awesome @mathiasvr video samples 😉 it seems that `HTMLMediaElement.play() Promise` is not as fine grained as the previous implementation 😭
> 
> So unless someone comes with a better solution, I think that the only proper option is to try to optimize the previous `onCanPlay` function 😅

Agreed :) +1

**Comment by mathiasvr** [comment]
@hicom150 Did you see other problems than muted audio? Otherwise I think the `play() Promise` could still be a good solution, if we can detect the audio problem (maybe using canPlay) as well.

**Comment by hicom150** [comment]
I changed the approach and instead of using  `HTMLMediaElement.play() Promise` I try with experimental `videoTracks` and `audioTracks` which describe the number of available (correctly decoded) tracks.

Having this, I see two options:

- If we use the same approach we use before, we can say that for a `video` element we expect at least one video track and one audio track and for an `audio` element we expect at least one audio track. (this has the caveat that videos with no audio track are treated like error #759)

- The other option could be to allow to play videos without any audio track either because it has not any audio stream or because the audio stream codec is not supported. (this has the caveat that we directly play videos without supported audio without giving the option to open in VLC)

To sum up, the advantage of this method is that we can check available tracks early in `onLoadedMetadata` event and that we don't rely on `webkitVideoDecodedByteCount` and `webkitAudioDecodedByteCount` that sometimes have timing issues.

**Comment by mathiasvr** [comment]
Nice work @hicom150!
I haven’t tested it yet, but definitely sounds more stable than the current approach.
I think not supporting video with no audio is fine for now, we could consider adding a “play without audio” button on the “Audio codec not supported” error dialog? But maybe it’s a little too hacky 😉

**Comment by RecoX** [comment]
> Nice work @hicom150!
> I haven’t tested it yet, but definitely sounds more stable than the current approach.
> I think not supporting video with no audio is fine for now, we could consider adding a “play without audio” button on the “Audio codec not supported” error dialog? But maybe it’s a little too hacky 😉

I feel its too hacky too. If we can't play the video as we should we simply can offer to open it with a an external player and that's it. Keep it simple.

**Comment by hicom150** [comment]
After some back and forth I think that finally I'm going to merge this PR a go forward 😉
Thank you for your awesome help in this issue guys! 💪
